things/model/src/main/java/org/eclipse/ditto/things/model/signals/commands/modify/MigrateThingDefinition.java

```diff
@@ -0,0 +1,312 @@
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import javax.annotation.concurrent.Immutable;
+
+import org.eclipse.ditto.base.model.headers.DittoHeaders;
+import org.eclipse.ditto.base.model.json.FieldType;
+import org.eclipse.ditto.base.model.json.JsonParsableCommand;
+import org.eclipse.ditto.base.model.json.JsonSchemaVersion;
+import org.eclipse.ditto.base.model.signals.FeatureToggle;
+import org.eclipse.ditto.base.model.signals.UnsupportedSchemaVersionException;
+import org.eclipse.ditto.base.model.signals.commands.AbstractCommand;
+import org.eclipse.ditto.json.JsonFactory;
+import org.eclipse.ditto.json.JsonField;
+import org.eclipse.ditto.json.JsonFieldDefinition;
+import org.eclipse.ditto.json.JsonObject;
+import org.eclipse.ditto.json.JsonObjectBuilder;
+import org.eclipse.ditto.json.JsonPointer;
+import org.eclipse.ditto.json.JsonValue;
+import org.eclipse.ditto.policies.model.ResourceKey;
+import org.eclipse.ditto.things.model.ThingId;
+import org.eclipse.ditto.things.model.signals.commands.ThingCommand;
+import org.eclipse.ditto.things.model.signals.commands.ThingCommandSizeValidator;
+import org.eclipse.ditto.things.model.signals.commands.exceptions.ThingIdNotExplicitlySettableException;
+
+/**
+ * Command to update a Thing's definition and apply a migration payload.
+ */
+@Immutable
+@JsonParsableCommand(typePrefix = ThingCommand.TYPE_PREFIX, name = MigrateThingDefinition.NAME)
+public final class MigrateThingDefinition extends AbstractCommand<MigrateThingDefinition>
+        implements ThingModifyCommand<MigrateThingDefinition> {
+
+    /**
+     * Name of the "Update Thing Definition" command.
+     */
+    public static final String NAME = "migrateThingDefinition";
+
+    /**
+     * Type of this command.
+     */
+    public static final String TYPE = TYPE_PREFIX + NAME;
+
+    private final ThingId thingId;
+    private final String thingDefinitionUrl;
+    private final JsonObject migrationPayload;
+    private final Map<ResourceKey, String> patchConditions;
+    private final Boolean initializeMissingPropertiesFromDefaults;
+
+    private MigrateThingDefinition(final ThingId thingId,
+            final String thingDefinitionUrl,
+            final JsonObject migrationPayload,
+            final Map<ResourceKey, String> patchConditions,
+            final Boolean initializeMissingPropertiesFromDefaults,
+            final DittoHeaders dittoHeaders) {
+        super(TYPE, FeatureToggle.checkMergeFeatureEnabled(TYPE, dittoHeaders));
+        this.thingId = checkNotNull(thingId, "thingId");
+        this.thingDefinitionUrl = checkNotNull(thingDefinitionUrl, "thingDefinitionUrl");
+        this.migrationPayload = checkJsonSize(checkNotNull(migrationPayload, "migrationPayload"), dittoHeaders);
+        this.patchConditions =
+                Collections.unmodifiableMap(patchConditions != null ? patchConditions : Collections.emptyMap());
```

Comment:
this parameter `patchConditions` is not marked as `@Nullable` and (I think) cannot be `null` from the code flow.  
So no need to check for null ..